Basics of 'subid ranges' support for IPA provider

alexey-tikhonov opened the pull request:
This is for preliminary review.

Limitations:
  - only IPA provider
  - single subid interval pair (subuid+subgid) per user
  - idviews aren't supported
  - only forward lookup (user -> subid)
    
Known TODOs:
  - delete cached subid ranges in case "not found" on a server
  - distinguish "user not found" vs "user doesn't have ranges defined"

To test:
  - build shadow-utils from latest upstream (or get Fedora Rawhide build from https://copr.fedorainfracloud.org/coprs/ipedrosa/subid_ranges/)  [Update: now available in Fedora Rawhide starting `shadow-utils-4.8.1-11.fc35`]
  - build FreeIPA with https://github.com/freeipa/freeipa/pull/5713 (or ping me for details of internal test setup)
  - build SSSD with this PR (enabling `--with-subid`)
  - edit /etc/nsswitch.conf: `subid: sss`
  - use for example this [utility](https://github.com/shadow-maint/shadow/blob/master/src/list_subid_ranges.c) from shadow-utils upstream to fetch subid ranges from IPA server (or `new?idmap`)

ikerexxe commented on src/sss_client/subid/sss_subid.c:
```diff
@@ -0,0 +1,206 @@
+#include "sss_cli.h"
+
+
+/* Find all subid ranges delegated to a user.
```
Can you use doxygen format to comment the APIs? Example: https://github.com/SSSD/sssd/blob/89a40e77a1477a3957f4ddc47890eaecbc4d5c7c/src/confdb/confdb.c#L2530

I think it only applies to the three functions in this file but I might be mistaken.

alexey-tikhonov replied:
This is an implementation of https://github.com/shadow-maint/shadow/blob/d4b6d1549b2af48ce3cb6ff78d9892095fb8fdd9/lib/prototypes.h#L271
Does it make sense to copy&paste doxygen comments from original API definition?

pbrezina replied:
No, but it might be nice to provide link to this doc.

alexey-tikhonov replied:
Fixed.

ikerexxe commented on src/providers/ldap/ldap_id_subid.c:
```diff
@@ -0,0 +1,264 @@
+        DEBUG(SSSDBG_MINOR_FAILURE,
+              "No such user '%s' or user doesn't have subid range\n",
+              state->name);
+        /* TODO: delete cached ranges? */
```
Did you leave this comment on purpose?

alexey-tikhonov replied:
Fixed.

ikerexxe reviewed (commented):
I've reviewed the changes and I have found several minor concerns. In any case, the scope of the changes is too big for me so I recommend that you get the feedback from a senior member of the team.

pbrezina commented on src/db/sysdb.h:
```diff
@@ -143,6 +143,12 @@
+#define SYSDB_SUBID_GID_COUNT   "subGidCount"
+#define SYSDB_SUBID_UID_NUMBER  "subUidNumber"
+#define SYSDB_SUBID_GID_NUMBER  "subGidNumber"
+#define SYSDB_SUBID_OWNER       "owner"
```
subOwner?

alexey-tikhonov replied:
Fixed.

pbrezina commented on src/db/sysdb_subid.c:
```diff
@@ -0,0 +1,158 @@
+
+errno_t sysdb_store_subid_range(struct sss_domain_info *domain,
+                                const char *name,
+                                int cache_timeout,
```
expiration_time?

alexey-tikhonov replied:
> * Please, provide some description to the first commit message.

Do you think it makes sense to keep 2 commits? I was going to squash them after review.


> * Why conditional build? Why is build disabled by default?

This is MVP of experimental feature.
Corresponding code in shadow-utils is merged upstream, but there is no upstream release available yet. We brought functionality to Fedora via patches. But I doubt other distributions will do the same. I.e. out of Fedora/RHEL, this is at the moment for enthusiasts who are willing to play with a new feature and who is capable to rebuild shadow-utils from sources.
Moreover, I anticipate significant changes might be required once we have initial feedback.

So... at the very least it's conditional to allow distributions to avoid installing plugin that they can't use yet.

Once shadow-utils upstream release is done and widely used, Podman patches released (https://github.com/containers/storage/pull/882), etc (so it is more or less easy to use), we can change default and build code / install plugin in sssd-ipa, IMO.

>     * We don't use two blank lines between function definitions.

And this (monolithic text) makes me sad :) I didn't see such restrictions in https://sssd.io/contrib/coding-style.html


>     * Can the range be set also for trusted domain user? If yes, this implementation doesn't seem to support it.

No, FreeIPA doesn't support it in MVP.

alexey-tikhonov replied:
Fixed.

pbrezina commented on src/providers/ldap/ldap_id_subid.c:
```diff
@@ -0,0 +1,264 @@
+        return;
+    }
+
+    subreq = sdap_get_generic_send(state,
```
More then one search base can be defined, you need to iterate over all. Use `sdap_search_bases()`?

alexey-tikhonov replied:
If I understand correctly, design of current solution in FreeIPA doesn't support multiple bases, so supporting this in SSSD doesn't make a lot of sense.

But formally speaking you are correct, of course. I just was lost in sdap code and tried to implement minimal operable solution (as I wasn't sure approach is viable at all).

I'll try to expand this now to support multiple bases.

alexey-tikhonov replied:
In particular, I was looking at "deskprofile" that seems to expect single search base specified:
https://github.com/SSSD/sssd/blob/7fdff741a7acaf016390988e5795df6bf03e7605/src/providers/ipa/ipa_deskprofile_config.c#L64

alexey-tikhonov replied:
I took `sdap_async_sudo` as an example and added one patch to make use of `sdap_search_bases`.
(Patch to be squashed later.)

pbrezina commented on src/providers/ldap/ldap_id_subid.c:
```diff
@@ -0,0 +1,264 @@
+
+        /* store range */
+        sysdb_store_subid_range(state->domain, state->name,
+                                state->domain->user_timeout,
```
Introduce its own entry_cache_subid_timeout options?

alexey-tikhonov replied:
User's subid range feels and looks like other NSS attributes. Does it really make sense to have different expiration time?

pbrezina replied:
subid range may be for user or group, right? We have expiration time for user and for group so it may be confusing that group subid expires with users.

alexey-tikhonov replied:
In general, `user` might have defined a set of `subuid` and/or `subgid` ranges(*) (sub**u**id - a range of subordinate UIDs, sub**g**is - a range of subordinate GIDs). Those ranges (both types) can be used by this user for mapping in parent namespace. So both types of subid ranges are kind of user's attribute.

`group` (as NSS object) can't have any subids defined.


(*) In current implementation FreeIPA supports only two options:
 - no subid ranges defined for user
 - user has both subuid and subgid defined (with the same fixed size 65k and the same interval start, but this implementation detail doesn't matter much for us)

pbrezina commented on src/providers/ldap/ldap_id_subid.c:
```diff
@@ -0,0 +1,264 @@
+        /* store range */
+        sysdb_store_subid_range(state->domain, state->name,
+                                state->domain->user_timeout,
+                                time(NULL), results[0]);
```
I think you can move the `now` parameter inside the function.

alexey-tikhonov replied:
Fixed.

pbrezina commented on src/sss_client/subid/sss_subid.c:
```diff
@@ -18,26 +18,111 @@
+ * 0-3: 32bit number with "start" id
+ * 4-7: 32bit number with "count" (range size)
  */
 enum subid_status shadow_subid_list_owner_ranges(const char *user,
```
This change seems to belong to the first patch.

alexey-tikhonov replied:
Actually I kept two patches to easy review a bit, but I was going to squash them after (and add a nice commit message). IMO, having commit with a stub of plugin in the history doesn't have much value.

pbrezina reviewed (commented):
Hi, I did not test it yet. See minor comments below and inline in the code.

- Please, provide some description to the first commit message.
- Why conditional build? Why is build disabled by default?
- We don't use two blank lines between function definitions.
- Can the range be set also for trusted domain user? If yes, this implementation doesn't seem to support it.

alexey-tikhonov commented:
Hi @pbrezina,

thank you for the review.

> * Please, provide some description to the first commit message.

Do you think it makes sense to keep 2 commits? I was going to squash them after review.


> * Why conditional build? Why is build disabled by default?

This is MVP of experimental feature.
Corresponding code in shadow-utils is merged upstream, but there is no upstream release available yet. We brought functionality to Fedora via patches. But I doubt other distributions will do the same. I.e. out of Fedora/RHEL, this is at the moment for enthusiasts who are willing to play with a new feature and who is capable to rebuild shadow-utils from sources.
Moreover, I anticipate significant changes might be required once we have initial feedback.

So... at the very least it's conditional to allow distributions to avoid installing plugin that they can't use yet.

Once shadow-utils upstream release is done and widely used, Podman patches released (https://github.com/containers/storage/pull/882), etc (so it is more or less easy to use), we can change default and build code / install plugin in sssd-ipa, IMO.

>     * We don't use two blank lines between function definitions.

And this (monolithic text) makes me sad :) I didn't see such restrictions in https://sssd.io/contrib/coding-style.html


>     * Can the range be set also for trusted domain user? If yes, this implementation doesn't seem to support it.

No, FreeIPA doesn't support it in MVP.

alexey-tikhonov commented on src/sss_client/subid/sss_subid.c:
```diff
@@ -0,0 +1,206 @@
+                                         bool *result)
+{
+    enum subid_status ret;
+    struct subid_range *r;
```
Fixed.

alexey-tikhonov commented on src/sss_client/subid/sss_subid.c:
```diff
@@ -0,0 +1,206 @@
+     *
+     * It might have some sense to do a check at responder's side (i.e. without
+     * fetching all ranges), but range is just a couple of numbers (and FreeIPA
+     * only support a single range per user anyway), so this optimization
```
Fixed.

alexey-tikhonov commented:
`make-check-valgrind` times out on Rawhide (and a couple of tests fail) but I can't reproduce this locally.

pbrezina commented:
> Hi @pbrezina,
> 
> thank you for the review.
> 
> > * Please, provide some description to the first commit message.
> 
> Do you think it makes sense to keep 2 commits? I was going to squash them after review.

It's up to you, just make sure it has sufficient description.

> > * Why conditional build? Why is build disabled by default?
> 
> This is MVP of experimental feature.
> Corresponding code in shadow-utils is merged upstream, but there is no upstream release available yet. We brought functionality to Fedora via patches. But I doubt other distributions will do the same. I.e. out of Fedora/RHEL, this is at the moment for enthusiasts who are willing to play with a new feature and who is capable to rebuild shadow-utils from sources.
> Moreover, I anticipate significant changes might be required once we have initial feedback.
> 
> So... at the very least it's conditional to allow distributions to avoid installing plugin that they can't use yet.

Ack. This is something to document in the commit message.

> Once shadow-utils upstream release is done and widely used, Podman patches released ([containers/storage#882](https://github.com/containers/storage/pull/882)), etc (so it is more or less easy to use), we can change default and build code / install plugin in sssd-ipa, IMO.

Ok.

> > ```
> > * We don't use two blank lines between function definitions.
> > ```
> 
> And this (monolithic text) makes me sad :) I didn't see such restrictions in https://sssd.io/contrib/coding-style.html

:-) Even though something is not explicitly banned, it is not cool to differentiate from the rest of the project.

> > ```
> > * Can the range be set also for trusted domain user? If yes, this implementation doesn't seem to support it.
> > ```
> 
> No, FreeIPA doesn't support it in MVP.

Code looks good to me. I'm going to test it.

pbrezina commented:
Seems to work fine. Please, do the final touch and I wlll ack it.

alexey-tikhonov commented:
Squashed patches, removed excessive new lines, added release notes to the commit message.

pbrezina commented:
> Squashed patches, removed excessive new lines, added release notes to the commit message.
> 
> Should I reference [shadow-maint/shadow#154](https://github.com/shadow-maint/shadow/issues/154) explicitly in the commit message?

It might be good for other distro maintainers.

Please separate the feature release note with blank line from the rest of description.

alexey-tikhonov commented:
> > Squashed patches, removed excessive new lines, added release notes to the commit message.
> > Should I reference [shadow-maint/shadow#154](https://github.com/shadow-maint/shadow/issues/154) explicitly in the commit message?
> 
> It might be good for other distro maintainers.

Done.

> Please separate the feature release note with blank line from the rest of description.

Well, I meant all this text to be included in the notes...

pbrezina commented:
Pushed PR: https://github.com/SSSD/sssd/pull/5693

* `master`
    * f546088226872f24722bdd94388816792bd5891a - Basics of 'subid ranges' support for IPA provider.
